### ISSUE TITLE
How to handle multiple widgets with same "implicit" ID

### ISSUE THREAD

**Issue by swistakm** [swistakm/pyimgui#56]
This isn't actually a bug but it may look like one.

Some widgets get same "implicit" id and usually this is the value of widget's label parameter. For instance:

```
if imgui.button("OK"):
    print("OK1")
if imgui.button("OK"):
    print("OK2")
```
All above buttons will have same label `"OK"`. Both buttons will look like clickable but only the first one will trigger on clicking. This would be a problem for common labels like yes/no/OK/cancel etc. but these usually are on different windows and windows have different IDs scopes. e.g. if we use following code:

```
imgui.begin("Win1")
if imgui.button("OK"):
    print("W1-OK1")
if imgui.button("OK"):
    print("W1-OK2")
imgui.end()

imgui.begin("Win1")
if imgui.button("OK"):
    print("W2-OK1")
if imgui.button("OK"):
    print("W2-OK2")
imgui.end()
```
We will see `W1-OK1` and `W2-OK1` in output but no `W1-OK2` nor `W2-OK2`. ImGui exposes following functions to maintain such ID scopes:
```
    void PushID(const char* str_id)
    void PushID(const char* str_id_begin, const char* str_id_end)
    void PushID(const void* ptr_id)
    void PushID(int int_id)
    void PopID()
```

In order to solve such problems they need to be ported. We already have additional context managers for managing push/pop actions on fonts and styles so we could also create custom `scope` or `scoped` context manager that would make code more readable. We would then have two ways to interact with scopes:

```
if imgui.button("OK"):
    print("OK outside scope")

imgui.push_id("my scope")
if imgui.button("OK"):
    print("OK in scope")
imgui.pop_id()

# or ...

if imgui.button("OK"):
    print("OK outside scope")

with imgui.scope("my scope"):
    if imgui.button("OK"):
        print("OK in scope")
```

**Comment by overdev** [comment]
Off-topic: Good to have you back. I was beginning to think you gave up on pyimgui.

This happens only in pyimgui? I mean, it does not occur in C++'s imgui?

**Comment by swistakm** [comment]
Oh, I didn't gave up on this project :). I simply had a very intense period at work so had to put few open source projects on hiatus. But I'm trying to get back to them.

Regarding the issue. I'm pretty sure this happens in barre C++ ImGui too because the this is the way how most immediate-mode GUIs are implemented i.e. every widget get's it's unique and (usually) explicit identifier. In this case the labels are used as identifiers. Usually this is acceptable simplification because it isn't common to have multiple same-labelled widgets in the same context (window, modal etc.)

**Comment by pconerly** [comment]
I like the idea of `imgui.scope`, but I wonder if we should put it into a different namespace? Having the python `imgui` module be a combination of `imgui-cpp` and `pyimgui` helpers might be confusing for coders?

Maybe `imgui.helpers.scope` ? or `imgui.py_scope`? I don't know; naming is hard.

**Comment by swistakm** [comment]
It is definitely a good idea to have the extra functions and utilities that do not exist in C++ API in separate module. It will improve code organization and help in documenting it properly. We already have few of such functions e.g. `font`, `styled`, `istyled`, `vertex_buffer_*`.

Still I would not care much about developer confusion at this moment. I think that benefits of having them importable from one place instead of multiple submodules outweights the cons. This is also in spirit of the one-header libraries like ImGui itself.

My idea is to have at least two Cython modules:
- `core.pyx` - for everything that exists directly in core C++ API
- `extras.pyx` - for anything extra (python-specific helpers, context managers) that is not part of C++ API

But combine them in `__init__.py` in following way:
```python
from imgui.core import *  # noqa
from imgui import core
from imgui.extras import *  # noqa
from imgui import extras
```

That way we will help to maintain backwards compatibility, make possible for some developers to be very specific about core/extras usage, and also make possible for other developers to have very simpe one-import code.

**Comment by 602p** [comment]
I've submitted a PR for a patch to expose PushID and PopID and a context wrapper as discussed. 

It's also possible to append some unique text to a name after `##` to make ids not collide. E.g. `imgui.button('foo##1')` and `imgui.button('foo##2')` don't collide, but are both displayed as foo.

**Comment by swistakm** [comment]
Closing this as we already have `imgui.extra` submodule and I have added new `scoped` context manager.

**Comment by swistakm** [comment]
Feature available in 1.2.0.
